### ISSUE TITLE
Implement buffered io streams (io.TextIOWrapper and friends)

### ISSUE THREAD

**Issue by pfalcon** [micropython/micropython#726]
Doing I/O (well, at least input) with multibyte characters (like UTF-8) requires buffering support (because if have bytes not forming a complete character, they cannot be passed to the application, buy yet we already have them, and need to store somewhere in the interim - in the buffer).

With this blocking need for input buffering to implement unicode text streams, Python "io" module actually defines buffered streams for binary data either, and buffered streams are used by default.

General buffering support is contradictory question. On one hand, uPy is intended to be light and unbloated, so buffering is rather left to underlying layers, like OS. But then one big usecase for uPy is baremetal, where no OS exists. But then it's particular driver's job to do the buffering, because disk/network drivers need to do it anyway. But that buffering is very likely the minimal required to accommodate hardware requirements, not giving the best performance (which is the main reason for application-level buffering).

Summing up, it makes sense to implement just generic I/O buffering in uPy, and give people choice to use it, while setting reasonable defaults.

**Comment by jepler** [comment]
There's now something _called_ `TextIOWrapper` but it is not compatible with Python:
```
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
>>> bio = io.BytesIO(b'bob')
>>> tio = io.TextIOWrapper(bio)
>>> tio.read()
'bob'
```

MicroPython:
```
MicroPython v1.18-244-g66fe3d5cb on 2022-06-01; linux [GCC 10.2.1] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import io
>>> bio = io.BytesIO(b'bob')
>>> io.TextIOWrapper(bio)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't convert 'BytesIO' object to str implicitly
```

This came up because I was investigating adding a json walker/visitor based on [json-stream](https://github.com/daggaz/json-stream#visitor-pattern) and it uses this idiom (with an encoding= kwarg) to convert a bytes-based stream to a str-based one.

**Comment by dpgeorge** [comment]
@jepler yes you're right, `io.TextTIOWrapper` is not fully compatible with CPython.  I wonder if it has any compatibility with CPython at all in its current form?  There are no tests for this class.

**Comment by dpgeorge** [comment]
`io.TextIOWrapper` was removed in e65d1e69e88268145ff0e7e73240f028885915be

I'll now close this issue (although it may be interesting to one day implement `io.TextIOWrapper` to match CPython).
